sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/MergingSessionsIterator.scala

```diff
@@ -0,0 +1,254 @@
+  private[this] var firstRowInNextGroup: InternalRow = _
+
+  // Indicates if we has new group of rows from the sorted input iterator
+  private[this] var sortedInputHasNewGroup: Boolean = false
+
+  // The aggregation buffer used by the sort-based aggregation.
+  private[this] val sortBasedAggregationBuffer: InternalRow = newBuffer
+
+  private[this] val groupingWithoutSessionProjection: UnsafeProjection =
+    UnsafeProjection.create(groupingWithoutSession, valueAttributes)
+
+  private[this] val sessionProjection: UnsafeProjection =
+    UnsafeProjection.create(Seq(sessionExpression), valueAttributes)
+
+  // The flag indicating there's error on iterator, sort precondition is not fulfilled.
+  private var errorOnIterator: Boolean = false
+
+  protected def initialize(): Unit = {
+    if (inputIterator.hasNext) {
+      initializeBuffer(sortBasedAggregationBuffer)
+      val inputRow = inputIterator.next()
+      nextGroupingKey = groupingWithoutSessionProjection(inputRow).copy()
+      val session = sessionProjection(inputRow)
+      nextGroupingSession = session.getStruct(0, 2).copy()
+      firstRowInNextGroup = inputRow.copy()
+      sortedInputHasNewGroup = true
+    } else {
+      // This inputIter is empty.
+      sortedInputHasNewGroup = false
+    }
+  }
+
+  initialize()
+
+  /** Processes rows in the current group. It will stop when it find a new group. */
+  protected def processCurrentSortedGroup(): Unit = {
+    currentGroupingKey = nextGroupingKey
+    currentSession = nextGroupingSession
+
+    // Now, we will start to find all rows belonging to this group.
+    // We create a variable to track if we see the next group.
+    var findNextPartition = false
+    // firstRowInNextGroup is the first row of this group. We first process it.
+    processRow(sortBasedAggregationBuffer, firstRowInNextGroup)
+
+    // The search will stop when we see the next group or there is no
+    // input row left in the iter.
+    while (!findNextPartition && inputIterator.hasNext) {
+      // Get the grouping key.
+      val currentRow = inputIterator.next()
+      val groupingKey = groupingWithoutSessionProjection(currentRow)
+
+      val session = sessionProjection(currentRow)
+      val sessionStruct = session.getStruct(0, 2)
+      val sessionStart = getSessionStart(sessionStruct)
+      val sessionEnd = getSessionEnd(sessionStruct)
+
+      // Check if the current row belongs the current input row.
+      if (currentGroupingKey == groupingKey) {
+        if (sessionStart < getSessionStart(currentSession)) {
+          errorOnIterator = true
+          throw new IllegalStateException("Input iterator is not sorted based on session!")
+        } else if (sessionStart <= getSessionEnd(currentSession)) {
+          // expanding session length if needed
+          expandEndOfCurrentSession(sessionEnd)
+          processRow(sortBasedAggregationBuffer, currentRow)
+        } else {
+          // We find a new group.
```

Comment:
To distinguish with line 169, maybe comment `we find a new session window for the same grouping key` here?